python/kwiver/vital/modules/module_helpers.cxx

```diff
@@ -0,0 +1,169 @@
+ *    this list of conditions and the following disclaimer.
+ *
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <python/kwiver/vital/modules/module_helpers.h>
+
+#include <python/kwiver/vital/util/python.h>
+#include <vital/plugin_loader/plugin_loader.h>
+#include <kwiversys/SystemTools.hxx>
+
+#include <pybind11/pybind11.h>
+
+#ifdef VITAL_LOAD_PYLIB_SYM
+  #include <dlfcn.h>
+#endif
+
+// Undefine macros that will double expand in case an definition has a value
+// something like: /usr/lib/x86_64-linux-gnu/libpython2.7.so
+#ifdef linux
+#define _orig_linux linux
+#undef linux
+#endif
+
+namespace py = pybind11;
+
+/*
+ * Check if a python interpretor exists and initialize one the python interpretor
+ * does not exist
+ */
+void check_and_initialize_python_interpretor()
+{
+  // Check if a python interpreter already exists so we don't clobber sys.argv
+  // (e.g. if sprokit is initialized from python)
+  if (!Py_IsInitialized())
+  {
+    // Embed a python interpretter if one does not exist
+    Py_Initialize();
+
+    // Set Python interpeter attribute: sys.argv = []
+    // parameters are: (argc, argv, updatepath)
+    PySys_SetArgvEx(0, NULL, 0);
+  }
+
+  if (!PyEval_ThreadsInitialized())
+  {
+    {
+      // Let pybind11 initialize threads and set up its internal data structures
+      pybind11::detail::get_internals();
+    }
+    // Release the GIL
+    PyEval_SaveThread();
+  }
+}
+
+/*
+ * check if the python shared library symbols have been loaded
+ * From: https://stackoverflow.com/questions/50570223/dlopenrtld-noload-still-returns-not-null-after-dlclose
+ */
+bool is_python_library_loaded(const std::string& python_library_path)
+{
+  void *handle = dlopen(python_library_path.c_str(), RTLD_NOW | RTLD_NOLOAD);
+  if ( handle != nullptr )
+  {
+    dlclose(handle);
```

Comment:
You can't use any of this on Windows. Use KWSys instead. I feel like we've discussed this one before. This code is causing the Windows dashboard failure.

Reply:
You are right @dstoup we have discussed this before on #925. Based on my understanding, dlopen code is not used in windows at all. This PR was branched of from #925 just to add logs. The [DVITAL_LOAD_PYLIB_SYM](https://github.com/Kitware/kwiver/blob/master/vital/bindings/python/vital/modules/CMakeLists.txt#L15) definition and [#ifdef VITAL_LOAD_PYLIB_SYM](https://github.com/Kitware/kwiver/blob/master/vital/bindings/python/vital/modules/registration.cxx#L137) are used in master and in #925 avoid the dlopen call.